### PR TITLE
Bugfix/isolation groups domain drains

### PR DESCRIPTION
**Pull request by davidporter-id-au**
**What changed?**
Fixes several bugs in the domain isolation-group handling which wasn't tested properly in replication it notably:
- Fixes the problem of upserting configuration from the inactive region, which was previously would error
- Fixes the problem of replication of configuration, which was entirely not working
- Refactors the domain controller by splitting out this functionality into its own, much simpler function rather than continuing to overload the already incomprehensible domain controller. 

**Why?**

**How did you test it?**
```
cadence --env docstore-prod11 --proxy_region dca admin isolation-groups get-domain --domain cadence-canary-global
Isolation Groups State
asdf5            Drained
asfd             Drained
-------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » cadence --env an-env --proxy_region dca admin isolation-groups update-domain --domain cadence-canary-global  --remove-all-drains
-------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » cadence --env an-env --proxy_region dca admin isolation-groups get-domain --domain cadence-canary-global
-- No groups found --
-------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » cadence --env an-env --proxy_region phx admin isolation-groups get-domain --domain cadence-canary-global
-- No groups found --
```

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/domain/replicationTaskExecutor.go [hunk 1]
```diff
@@ -192,12 +194,14 @@ func (h *domainReplicationTaskExecutorImpl) handleDomainUpdateReplicationTask(ct
 	// task already validated
 	status, err := h.convertDomainStatusFromThrift(task.Info.Status)
 	if err != nil {
+		h.logger.Info("Error handling replication task", tag.Dynamic("task", task), tag.Error(err))
```
Comment (Shaddoll) [inline common/domain/replicationTaskExecutor.go:197]: we should avoid using Dynamic

File: common/domain/replicationTaskExecutor.go [hunk 2]
```diff
@@ -262,8 +268,10 @@ func (h *domainReplicationTaskExecutorImpl) handleDomainUpdateReplicationTask(ct
 	}
 
 	if !recordUpdated {
+		h.logger.Warn("Domain update failed, record not updated", tag.Dynamic("replication-request", request))
```
Comment (Shaddoll) [inline common/domain/replicationTaskExecutor.go:271]: this is an expected path, the log should be removed

File: common/domain/handler.go [hunk 3]
```diff
@@ -669,6 +658,111 @@ func (d *handlerImpl) DeprecateDomain(
...
+// UpdateIsolationGroups is used for draining and undraining of isolation-groups for a domain.
+// Like the isolation-group API, this controller expects Upsert semantics for
+// isolation-groups and does not modify any other domain information.
+//
+// Isolation-groups are regional in their configuration scope, so it's expected that this upsert
+// includes configuration for both clusters every time.
+//
+// The update is handled like other domain updates in that they expected to be replicated. So
+// unlike the global isolation-group API it shouldn't be necessary to call
+func (d *handlerImpl) UpdateIsolationGroups(
```
Comment (Shaddoll) [inline common/domain/handler.go:670]: The refactoring is not a good idea. It introduces duplicate logic and the most important thing is that replication is using `UpdateDomain`

File: common/domain/replicationTaskExecutor.go (hunk 2 above)
Comment (davidporter-id-au) [reply common/domain/replicationTaskExecutor.go:271]: Is.. it? I'd actually be surprised to see conditional failures like this

Comment (davidporter-id-au) [reply common/domain/replicationTaskExecutor.go:270]: Yes, I understand that, but what I was meaning to say was that conditional update failures are surprising, probably shouldn't happen unless something's not getting processed, unless I'm confused? 

You cannot trigger a domain update more than 30 seconds apart, so actually hitting this codepath seems... surprising

File: common/domain/handler.go [hunk 4]
```diff
@@ -669,6 +658,115 @@ func (d *handlerImpl) DeprecateDomain(
...
+	if lastUpdatedTime.Add(d.config.FailoverCoolDown(currentDomainConfig.Info.Name)).After(d.timeSource.Now()) {
+		return errDomainUpdateTooFrequent
+	}
+
+	// Check the failover cool down time
+	if lastUpdatedTime.Add(d.config.FailoverCoolDown(currentDomainConfig.Info.Name)).After(d.timeSource.Now()) {
+		return errDomainUpdateTooFrequent
+	}
+
+	if !d.clusterMetadata.IsPrimaryCluster() {
```
[the remark this answers is not in the document]
Comment (Shaddoll) [reply common/domain/handler.go:709]: only check for global domain

File: common/domain/replicationTaskExecutor.go [hunk 5]
```diff
@@ -262,6 +267,7 @@ func (h *domainReplicationTaskExecutorImpl) handleDomainUpdateReplicationTask(ct
 	}
 
 	if !recordUpdated {
+		h.logger.Warn("Domain update failed, record not updated")
```
Comment (Shaddoll) [inline common/domain/replicationTaskExecutor.go:270]: i still don't think we should add this log

Comment (davidporter-id-au) [reply common/domain/replicationTaskExecutor.go:270]: What is your concern? I don't see how this is a noisy log

Comment (Shaddoll) [reply common/domain/replicationTaskExecutor.go:270]: The message is misleading and it's asking people to investigate a non-issue.

Comment (davidporter-id-au) [reply common/domain/replicationTaskExecutor.go:270]: Maybe my mental model is wrong, under what conditions should a conditional update normally occur in this codepath? I get domain update queue replays, but that seems fairly rare

Comment (Shaddoll) [reply common/domain/replicationTaskExecutor.go:270]: I don't know the details, but it probably could occur when merging dlq.

File: common/domain/handler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidporter-id-au) [reply common/domain/handler.go]: good call

File: common/domain/replicationTaskExecutor.go [hunk 6]
```diff
@@ -198,6 +200,7 @@ func (h *domainReplicationTaskExecutorImpl) handleDomainUpdateReplicationTask(ct
 	// first we need to get the current notification version since we need to it for conditional update
 	metadata, err := h.domainManager.GetMetadata(ctx)
 	if err != nil {
+		h.logger.Info("Error getting metadata while handling replication task", tag.Error(err))
```
Comment (Shaddoll) [inline common/domain/replicationTaskExecutor.go:203]: nit: Error

**Review by Shaddoll — APPROVED** [review]

File: common/domain/replicationTaskExecutor.go (hunk 6 above)
Comment (davidporter-id-au) [reply common/domain/replicationTaskExecutor.go:203]: thank you

**Pull request merged**